docs: Generalize the data transformation of pipes (#37950)
angular/angular: aio/content/guide/pipes.md

# Transforming Data Using Pipes

> Use [pipes](guide/glossary#pipe "Definition of a pipe") to transform strings, currency amounts, dates, and other data for display (of contents, structure and styles etc.).

Reviewer: We're trying to avoid the use of parentheses as well as Latin abbreviations––for speakers of other languages with less Latin influence. I'd omit the parenthetical like this:
```suggestion
Use [pipes](guide/glossary#pipe "Definition of a pipe") to transform strings, currency amounts, dates, and other data for display.
```

> A pipe is a simple function you can use repeatedly in [template expressions](/guide/glossary#template-expression "Definition of template expression") to accept an input value and return a transformed value, without declaring it in multiple places.

Reviewer: I think you could leave out the word repeatedly because that is often you how you'd use a function. That is, ideally you wouldn't have many single-use functions. What about this?
```suggestion
A pipe is a function you can use in [template expressions](/guide/glossary#template-expression "Definition of template expression") to accept an input value and return a transformed value.
```
Notice I've removed "simple".

I'm not sure on the last phrase since this is the introductory paragraph on the topic. Getting into specific details like that might be better for later on somewhere else in the doc. @rockument69 might have some insights for us on why he chose the wording he did.

Reviewer 2: I think the introductory sentence should stay as it is now (recently updated): " Pipes are simple functions you can use in template expressions to accept an input value and return a transformed value." 

As for the initial suggestion, "and other data for display (of contents, structure and styles etc.)" I would argue that "transformation" covers all of this, including changing any property, structure, style, or content. It also includes translation, for example (which deals with content). 

Also, this part of the page is all about the built-in pipes. To do more than the built-in pipes do, you must create a custom pipe, and that section is later in the doc.

Reviewer 3: I think there's room for both ideas here.

The current introductory statement does a good job explaining what a pipe does. I think what @KingMario is trying to point out is that we could add more information about **why** you might want to use a pipe. Perhaps the following text might work?

```suggestion
Pipes are functions you can use in [template expressions](/guide/glossary#template-expression "Definition of template expression") to accept an input value and return a transformed value. Pipes are useful because you can use them throughout your application, while only declaring it once.
```

Author: Thanks @aikidave

Do you think the last sentence should be updated to `Pipes are useful because you can use them throughout your application, while only declaring each pipe once.`?

Reviewer 3: Excellent point! The word `it` can be very ambiguous. I agree with your suggestion!

# Transforming Data Using Pipes

Use [pipes](guide/glossary#pipe "Definition of a pipe") to transform strings, currency amounts, dates, and other data for display (of contents, structure and styles etc.).
A pipe is a simple function you can use repeatedly in [template expressions](/guide/glossary#template-expression "Definition of template expression") to accept an input value and return a transformed value, without declaring it in multiple places.
For example, you would use a pipe to show a date as **April 15, 1988** rather than the raw string format. 

> Another example is a pipe that you can use to check if a form control is touched and invalid. If so, you can modify the display style of the associated input element to highlight it.

Reviewer 4: I think this example is definitely valid, but feels be too specific for the guide that talks about Pipes in general.

Author: Thank you @AndrewKushnir.

Do you think I should remove this sample pipe, as for now, the idea that a pipe is not just for formatting has been clarified in the first paragraph?

Or may I shorten it like `You would also use a pipe to determine whether to highlight an input element by checking the status of the associated form control.`?

Thanks.

Reviewer 4: @KingMario I'd propose removing the sentence that starts with `Another example ...`. I think the date-based example is enough for this part of the document (more examples and information is available below).

Regarding other changes in the guide: I feel like current version is pretty accurate and I'd probably keep it as is for now. We could potentially improve it by re-phrasing/clarifying what "display data" is and which "raw string format" of the date we transform to **April 15, 1988**.

May be @kapunahelewong can provide additional input on this and whether we should change this guide.

Author: @AndrewKushnir I've removed the sentence that starts with `Another example...`.

Please review it again. Thanks.

{@a parameterizing-a-pipe}

> ## Transforming data with parameters and chained pipes

Reviewer: Is there a technical difference between formatting and transforming a string? I searched around and mostly found "formatting". Usually I found transforming in the case of CSS. It seems to me that there is probably a subtle difference, but I'm not sure what it is.

Author: Thanks @kapunahelewong 

I change this because, in my company, in a code review for a pipe, we had an argument that should a pipe only change the format of a data, or can it be used to extract the property or properties from a complex object.

The date format pipe sample and the title of this section given in this guide might make the readers think that pipes should only return formatted strings.

This is the pipe that raised the argument.

```typescript
@Pipe({
  name: 'touchedAndInvalid',
  pure: false,
})
export class TouchedAndInvalidPipe implements PipeTransform {
  transform(control: FormControl): boolean {
    return !!control && control.touched && control.invalid;
  }
}
```

Some of my peers think this should not be declared as a pipe because it doesn't return a formatted string, among them is one who had studied in the United States and has a good knowledge of English.

While according to @AndrewKushnir , it is definitely valid. 

A thought comes to me that this guide should also describe when to declare and use pipes.

I noticed, back in version 8 of this guide, there's a sentence `You may notice that you desire many of the same transformations repeatedly, both within and across many applications.`. I think it tells the essential of pipes, that is, you want to use it **repeatedly**, **directly** and **effortlessly** in template expressions without declaring it in every component.

As you have mentioned, `repeatedly` is often how you'd use a function. Then do you think I could add `directly` and `effortlessly` to the sentence like this:

`A pipe is a function you can use directly and effortlessly in [template expressions](/guide/glossary#template-expression "Definition of template expression") to accept an input value and return a transformed value.`

Does that make sense?

Thanks.

Reviewer 3: In this case, the use of the word `transform` is correct. The reason we're using the term is because pipes can do more than just format values; consider the `async` pipe, for example. We view formatting as a subset of transforming. 

I'm glad you brought this up, @KingMario! These discussions are really helpful for clarifying terminology.

Reviewer: Thank you both. I did not know the difference and now I do!

Reviewer: @KingMario, I think `repeatedly` might have been better in that it is more succinct. We definitely want to make sure that the sentence doesn't get too long.

Author: @kapunahelewong , thanks for the help from @aikidave , I think it's OK without `repeatedly` or `directly`, `effortlessly` now. The sentence added by @aikidave, `Pipes are useful because you can use them throughout your application, while only declaring it once.`, straightforwardly answered the question `when to declare and use pipes`.

And if you agree, I'll resolve this conversation regarding the term `transform`.

Thanks.
